docs/madr/decisions/053-sni.md

```diff
@@ -0,0 +1,60 @@
+# SNI for MeshService, MeshExternalService etc.
+
+* Status: accepted
+
+## Context and Problem Statement
+
+Currently, the SNI format is `<kuma.io/service>{mesh=<mesh>,tag1=value1}`. For example:
+* `backend{mesh=demo}`
+* `backend{mesh=demo,version=2}`
+
+We need to introduce SNI format for `MeshService` and `MeshExternalService` etc.
+Current format does not work, because we need to differentiate between `kuma.io/service` and MeshService.
+Current format also has other problems:
+* It violates [SNI spec](https://datatracker.ietf.org/doc/html/rfc3546#page-8). It should be a valid hostname. Using characters like `,{/` is not valid.
+  Envoy does not complain about it right now, but it can bite us in the future.
+* It can be too long. See https://github.com/kumahq/kuma/issues/8915
+
+## Considered Options
+
+* New format
+
+## Decision Outcome
+
+Chosen option: "New format", because it solves existing problem and lets us differentiate MeshService from old service.
+
+## Pros and Cons of the Options
+
+### New format
+
+It's important to note that we never parse SNI. We only use it for multizone communication on zone proxies to associate to which service pass the data.
+It means that we can hash the data as long as it's consistent hashing.
+We cannot put tags in the original form because they can violate hostname requirements. For example `/` in `kuma.io/service` is not ok.
+The new format could look like this then:
+`<resource-name>.<mesh-name>.<extra-hash-if-needed>.<resource-type>`
+* `<resource-name>` - (required) name of the resource. Name of MeshService or MeshExternalService
+* `<mesh-name>` - (required) name of the mesh
+* `<extra-hash-if-needed>` - (optional) 16 characters of SHA256. In case of MeshService tags it's `SHA256(<tag1>=<value1>,<tag2>=<value2>)` sorted by keys.
+* `<resource-type>` - (required) `ms` (MeshService) or `mes` (MeshExternalService)
+
+Examples:
+* `backend.demo.ms` - MeshService named `backend` in `demo` mesh
+* `redis.kuma-demo.default.ms` - MeshService named `redis` in `kuma-demo` namespace in `default` mesh
+* `google.default.mes` - MeshExternalService named `google` in `default` mesh
+* `backend.demo.e3b0c44298fc1c.ms` - subset of MeshService named `backend` in `demo` mesh for tags `version=v2`
+
+Resource name can be at most 255 chars long. Mesh should not be longer than 64 chars.
+If we go over the limit we compute SHA256 hash `<resource-name>.<mesh-name>.<extra-hash-if-needed>`.
+For example:
+* `super-long-name-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.default.e3b0c44298fc1c14.ms`
+  Changes to `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855.ms`
```

Comment:
I think @lobkovilya for resource sync did something that kept the beginning of the name. I think it's useful for troubleshooting as it keeps most of the info available.

Something like:

`<resource-name>.<mesh-name>.<extra-hash-if-needed>.<resource-type>`

If resource-name is too long we switch to: `<resource-name>[:16]hash(<resource-name)[:16]`

Thinking about it I think it would be beneficial to just define this format with a character capacity for each element.

.e.g:

Format is `<format>.<resource-name>.<resourceHash>.<mesh-name>.<hash>.<resourceType>`

We have 256-5 (separators) = 251 chars available.

- format is 1 character (helps us with differnt versions).
- resourceNameHash is at 16 charts
- resourceName is at most 151 chars if it's larger it's truncated
- meshName is at most 64 chars
- hash is 16 chars
- resourceType is at most 3 chars

Otherwise we route just by hash and the resourceName is only an indicator.
Format is: `<format>.<serviceHash>.<routeHash>.<resourceType>.<resourceName>.<meshName>`

- format is 1 char (versioning)
- serviceHash 16 first chars of  `sha256(<resourceName>,<meshName>)`
- resourceType is at most 3 chars
- route hash 16 first chars of  `sha256(labels...)`
- resourceName is at most 64 chars (if too long it's truncated)
- meshName is at most 64 chars

This way jsut the format and serviceHash and resourceType is useful, the rest is for debugging no?

Reply:
that works I think. The only question is do we include serviceHash always or only if we go over char limit. I think including it always might be a bit redundant, no?

Reply:
Is there a benefit of having separate hashes for each component? Can we have a single hash at the end but separate limits:
`format(3-bytes).resource-name(N-bytes).mesh-name(64-bytes).resource-type(K-bytes).hash(resource-name, mesh-name, resource-type, tags)`?

Reply:
> kept the beginning of the name. I think it's useful for troubleshooting as it keeps most of the info available.

➕

Reply:
Yes, we can have a single hash. Again - the question is. Do we always add it or on demand?
```
v1.backend.demo.ms
```
is cleaner and has less data than
```
v1.backend.demo.e3b0c44298fc1c14.ms
```

btw. I'd put resource type at the end, it's safer to get valid domain. hash can technically produce only numbers which is not valid TLD.

Any opinions?

Reply:
> Any opinions?

```
v1.backend.demo.e3b0c44298fc1c14.ms
```

visually I like

```
e3b0c44298fc1c14.v1.backend.demo.ms
```

more, because it looks more similar to `pkc-41231jofsda.us-east-2.aws-cloud` and other dynamically generated domains.

> Do we always add it or on demand?

I'd add it always - I value consistency and there won't be any confusion on "why does this have a hash and that does not?"

Reply:
I don't really have preferences between including or not including hash if components are not trimmed. I don't see pros or cons of either approach.

Though I'd avoid having format version as obvious as `v1.` as it might be easily assumed as version v1 of `backend`. Maybe version should be just a first character in the hash:
```
backend.demo.aXXXXXXXXXXXXXXXX.ms
```
26 lowercase letters + 10 digits should be enough to encode versions for many years ahead :)

Reply:
So what's the conclusion here?

Reply:
updated. Let me know if that's good.

I put version "into" hash, but I moved this to the beginning because I don't want to parse everything to guess which format it is.